source/slang/ir-bind-existentials.cpp

```diff
@@ -0,0 +1,267 @@
+
+                UInt firstSlot = resInfo->index;
+                UInt slotCount = 0;
+                if(auto typeResInfo = varLayout->getTypeLayout()->FindResourceInfo(LayoutResourceKind::ExistentialSlot))
+                    slotCount = UInt(typeResInfo->count.getFiniteValue());
+
+                // Okay, let's check if we have enough arguments.
+
+                if( !bindEntryPointExistentialSlotsInst )
+                {
+                    sink->diagnose(inst->sourceLoc, Diagnostics::unexpected, "no entry point existential bindings");
+                    break;
+                }
+
+                auto operandsForInst = bindEntryPointExistentialSlotsInst->getOperands() + firstSlot;
+                replaceTypeUsingExistentialSlots(
+                    param,
+                    slotCount,
+                    operandsForInst);
+
+#if 0
+                UInt slotArgCount = bindEntryPointExistentialSlotsInst->getOperandCount();
+                if( (firstSlot + slotCount) > slotArgCount )
+                {
+                    sink->diagnose(inst->sourceLoc, Diagnostics::unexpected, "not enough global existential args");
+                    break;
+                }
+
+                // Okay, we can find just the operands we need.
+                List<IRInst*> argsForInst;
+                for( UInt ii = 0; ii < slotCount; ++ii )
+                {
+                    auto arg = bindEntryPointExistentialSlotsInst->getOperand(firstSlot + ii);
+                    argsForInst.Add(arg);
+                }
+
+                SharedIRBuilder sharedBuilder;
+                sharedBuilder.session = module->getSession();
+                sharedBuilder.module = module;
+
+                IRBuilder builder;
+                builder.sharedBuilder = &sharedBuilder;
+
+                // Move the decoration down to the parameter itself, so that we can manipulate it independently.
+                builder.addBindExistentialSlotsDecoration(param, argsForInst.Count(), argsForInst.Buffer());
+#endif
+            }
+
+            if( bindEntryPointExistentialSlotsInst )
+            {
+                bindEntryPointExistentialSlotsInst->removeAndDeallocate();
+            }
+        }
+
+        // Okay, now all the decorations are on the individual parameters... great...
+
+
+    }
+
+    void replaceTypeUsingExistentialSlots(
+        IRInst*         inst,
+        UInt            slotCount,
+        IRUse const*    slotArgs)
+    {
+        SLANG_UNUSED(slotCount);
+
+        auto fullType = inst->getFullType();
+        auto type = inst->getDataType();
+
+        SharedIRBuilder sharedBuilder;
+        sharedBuilder.session = module->getSession();
+        sharedBuilder.module = module;
+
+        IRBuilder builder;
+        builder.sharedBuilder = &sharedBuilder;
+
+        if( auto interfaceType = as<IRInterfaceType>(type) )
+        {
+            // This is the easy case...
+
+            auto newType = (IRType*) slotArgs[0].get();
```

Comment:
This is one of those tricky cases where I need an `IRType*` because that is what a function call later on needs, but the instruction there might not *actually* be derived from `IRType`.

As for the `slotArgs[0]` vs `slotArgs[1]` part: yeah, that probably needs to be documented better. The right long-term fix is to define an `IRInst` subclass that has accessors with the right semantics, but I don't want to codify things too much since I expect a lot of what is in this PR to get changed and refined as I continue the implementation, so baking in the details too much at this juncture might just make for more churn later.